shopify_function_macro/src/lib.rs

```diff
@@ -114,6 +114,150 @@ pub fn shopify_function(
+    };
+    let result = path.path.segments.last().unwrap();
+    assert!(
+        result.ident.to_string() == "Result",
+        "Shopify Functions must return a Result"
+    );
+    let PathArguments::AngleBracketed(generics) = &result.arguments else {
+        panic!("Shopify Function Result is missing generic arguments")
+    };
+    assert!(
+        generics.args.len() == 1,
+        "Shopify Function Result takes exactly one generic argument"
+    );
+    let GenericArgument::Type(ty) = generics.args.first().unwrap() else {
+        panic!("Shopify Function Result expects a type")
+    };
+    let Type::Path(path) = ty else {
+        panic!("Unexpected result type for Shopify Function Result")
+    };
+    &path.path.segments.last().as_ref().unwrap().ident
+}
+
+#[proc_macro_attribute]
+pub fn shopify_function_target(
+    attr: proc_macro::TokenStream,
+    item: proc_macro::TokenStream,
+) -> proc_macro::TokenStream {
+    let ast: syn::ItemFn = syn::parse(item).unwrap();
+    let args = parse_macro_input!(attr as ShopifyFunctionTargetArgs);
+
+    let name = &ast.sig.ident;
+    let name_as_stringlit: String = name.to_string();
+
+    let query_path = args.query_path.expect("No value given for query_path");
+    let schema_path = args.schema_path.expect("No value given for schema_path");
+    let output_query_file_name = format!(".{}{}", name_as_stringlit, OUTPUT_QUERY_FILE_NAME);
+
+    let input_struct = generate_struct(
+        "Input",
+        query_path.value().as_str(),
+        schema_path.value().as_str(),
+    );
+    let output_struct = generate_struct(
+        "Output",
+        &output_query_file_name,
+        schema_path.value().as_str(),
+    );
+    let output_result_type = extract_shopify_function_return_type(&ast)
+        .to_token_stream()
+        .to_string();
+    let output_query = format!(
+        "mutation Output($result: {}!) {{\n    {}(result: $result)\n}}\n",
+        output_result_type, name_as_stringlit
+    );
+
+    write_output_query_file(&output_query_file_name, &output_query);
+
+    let input_stream = args
+        .input_stream
+        .map_or(quote! { std::io::stdin() }, |stream| {
+            stream.to_token_stream()
+        });
+    let output_stream = args
+        .output_stream
+        .map_or(quote! { std::io::stdout() }, |stream| {
+            stream.to_token_stream()
+        });
+
+    quote! {
+        pub mod #name {
+            use super::*;
+            use std::io::Write;
+
+            #input_struct
+            #output_struct
+
+            #[shopify_function(
+                input_stream = #input_stream,
+                output_stream = #output_stream
+            )]
+            pub #ast
+
+            #[no_mangle]
+            #[export_name = #name_as_stringlit]
+            pub extern "C" fn export() {
+                main().unwrap();
+                #output_stream.flush().unwrap();
+            }
+        }
+    }
+    .into()
```

Comment:
I think I need this one?

```
error[E0308]: mismatched types
   --> shopify_function_macro/src/lib.rs:244:5
    |
201 |   ) -> proc_macro::TokenStream {
    |        ----------------------- expected `proc_macro::TokenStream` because of return type
...
244 | /     quote! {
245 | |         pub mod #export_ident {
246 | |             use super::*;
247 | |             use std::io::Write;
...   |
265 | |         #ast
266 | |     }
    | |_____^ expected `proc_macro::TokenStream`, found `TokenStream2`
    |
    = note: `TokenStream2` and `proc_macro::TokenStream` have similar names, but are actually distinct types
note: `TokenStream2` is defined in crate `proc_macro2`
   --> /Users/d/.cargo/registry/src/index.crates.io-6f17d22bba15001f/proc-macro2-1.0.46/src/lib.rs:161:1
    |
161 | pub struct TokenStream {
    | ^^^^^^^^^^^^^^^^^^^^^^
note: `proc_macro::TokenStream` is defined in crate `proc_macro`
   --> /Users/d/.rustup/toolchains/stable-aarch64-apple-darwin/lib/rustlib/src/rust/library/proc_macro/src/lib.rs:80:1
    |
80  | pub struct TokenStream(Option<bridge::client::TokenStream>);
    | ^^^^^^^^^^^^^^^^^^^^^^
    = note: this error originates in the macro `quote` (in Nightly builds, run with -Z macro-backtrace for more info)
help: call `Into::into` on this expression to convert `TokenStream2` into `proc_macro::TokenStream`
   --> /Users/d/.cargo/registry/src/index.crates.io-6f17d22bba15001f/quote-1.0.21/src/lib.rs:510:11
    |
510 |         _s.into()
    |           +++++++
```